config/file.go

```diff
@@ -0,0 +1,323 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package config
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils/fileutils"
+)
+
+var (
+	ReadOnlyConfigurationError = errors.New("configuration is read-only")
+)
+
+// fileStore is a config store backed by a file such as config/config.json.
+type fileStore struct {
+	emitter
+
+	config               *model.Config
+	environmentOverrides map[string]interface{}
+	configLock           sync.RWMutex
+	path                 string
+	watch                bool
+	watcher              *watcher
+	needsSave            bool
+}
+
+// NewFileStore creates a new instance of a config store backed by the given file path.
+//
+// If watch is true, any external changes to the file will force a reload.
+func NewFileStore(path string, watch bool) (fs *fileStore, needsSave bool, err error) {
+	resolvedPath, err := resolveConfigFilePath(path)
+	if err != nil {
+		return nil, false, err
+	}
+
+	fs = &fileStore{
+		path:  resolvedPath,
+		watch: watch,
+	}
+	if _, err = fs.Load(); err != nil {
+		return nil, false, err
+	}
+
+	if fs.watch {
+		if err = fs.startWatcher(); err != nil {
+			mlog.Error("failed to start config watcher", mlog.String("path", path), mlog.Err(err))
+		}
+	}
+
+	return fs, fs.needsSave, nil
+}
+
+// resolveConfigFilePath attempts to resolve the given configuration file path to an absolute path.
+//
+// Consideration is given to maintaining backwards compatibility when resolving the path to the
+// configuration file.
+func resolveConfigFilePath(path string) (string, error) {
```

Comment:
I'd love to deprecate relative config paths altogether if we preserve config files at all come v6.0.